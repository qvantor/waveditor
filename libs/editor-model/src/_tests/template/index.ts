export * as BasicCase from './basic-case';
export * as HostTemplate from './host-template';
