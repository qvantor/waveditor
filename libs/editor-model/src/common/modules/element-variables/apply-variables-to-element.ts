import {
  isTipTapVariable,
  mapTipTapParentsContent,
  TipTapVariable,
  TipTapVarLabel,
} from '@waveditors/utils';
import { JSONContent } from '@tiptap/core';
import {
  ElementCommon,
  ElementLinkUrl,
  getElementLinkUrl,
  isTextElement,
  Text,
  TextContent,
} from '../../../elements';
import { Variable, Variables } from '../../../variables';

const applyVariableToTipTapVariable = (
  variable: Variable,
  tipTapVar: TipTapVariable
): TipTapVariable | null => {
  if (variable.label === tipTapVar.attrs.label) return null;
  return TipTapVarLabel.set(variable.label)(tipTapVar);
};

const applyVariablesToJSONContent = (
  variables: Variables,
  content: JSONContent
): JSONContent | null => {
  let changed = false;
  const result = mapTipTapParentsContent(content, (items) =>
    items.reduce<JSONContent[]>((sum, item) => {
      if (!isTipTapVariable(item)) return [...sum, item];
      const variable = variables.find(
        (variable) => variable.id === item.attrs.id
      );

      // no variables mean that var was removed
      if (!variable) {
        changed = true;
        return sum;
      }
      const newItem = applyVariableToTipTapVariable(variable, item);
      if (!newItem) return [...sum, item];

      changed = true;
      return [...sum, newItem];
    }, [])
  );
  return changed ? result : null;
};

const applyVariablesToCommonElement = (
  variables: Variables,
  element: ElementCommon
): ElementCommon | null => {
  const url = getElementLinkUrl(element);
  if (!url || typeof url === 'string') return null;
  const newUrl = applyVariablesToJSONContent(variables, url);
  if (!newUrl) return null;
  return ElementLinkUrl.set(newUrl)(element);
};

const applyVariablesToText = (
  variables: Variables,
  text: Text
): Text | null => {
  const content = applyVariablesToJSONContent(variables, text.params.content);
  if (!content) return null;
  return TextContent.set(content)(text);
};

const applyVariablesByType = (
  variables: Variables,
  element: ElementCommon
): ElementCommon | null => {
  if (isTextElement(element)) return applyVariablesToText(variables, element);
  return null;
};

export const applyVariablesToElement = (
  variables: Variables,
  element: ElementCommon
) => {
  const common = applyVariablesToCommonElement(variables, element);
  const result = applyVariablesByType(variables, common ?? element);
  return result ?? common;
};
