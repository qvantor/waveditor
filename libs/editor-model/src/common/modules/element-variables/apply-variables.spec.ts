import { BasicCase } from '../../../_tests';
import { applyVariablesToElement } from './apply-variables-to-element';

const { Template, TextWithNameVariable, NameVariable } = BasicCase;

const TextWithNameVariableNode = Template.elements[TextWithNameVariable];

describe('apply variables', () => {
  it('should do nothing, if variables the same', () => {
    expect(
      applyVariablesToElement([NameVariable], TextWithNameVariableNode)
    ).toBeNull();
  });
  it('should change the var name in text content, if name is changed', () => {
    expect(
      applyVariablesToElement(
        [{ ...NameVariable, label: 'Hello world' }],
        TextWithNameVariableNode
      )
    ).toMatchSnapshot();
  });
  it('should remove variable, if variable gone', () => {
    expect(
      applyVariablesToElement([], TextWithNameVariableNode)
    ).toMatchSnapshot();
  });
});
