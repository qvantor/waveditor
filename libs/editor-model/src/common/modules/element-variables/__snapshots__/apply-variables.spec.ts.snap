// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply variables should change the var name in text content, if name is changed 1`] = `
Object {
  "id": "0.47976090123156934",
  "link": null,
  "name": "text1",
  "params": Object {
    "content": Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "hello ",
              "type": "text",
            },
            Object {
              "attrs": Object {
                "id": "0.7830225981794139",
                "label": "Hello world",
              },
              "type": "variable",
            },
            Object {
              "text": " ",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "doc",
    },
  },
  "style": Object {},
  "type": "text",
}
`;

exports[`apply variables should remove variable, if variable gone 1`] = `
Object {
  "id": "0.47976090123156934",
  "link": null,
  "name": "text1",
  "params": Object {
    "content": Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "hello ",
              "type": "text",
            },
            Object {
              "text": " ",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "doc",
    },
  },
  "style": Object {},
  "type": "text",
}
`;
